File: rust/question.rs
use std::io::{self, Write};

fn question(prompt: &str, valid: Option<&[&str]>) -> io::Result<String> {
    let joined_valid = valid.map(|v| v.join(", "));
    let mut input = String::new();

    loop {
        println!("{}", prompt);
        if let Some(ref valid) = joined_valid {
            print!("({})", valid);
        }
        print!(": ");

        io::stdout().flush()?;
        io::stdin().read_line(&mut input)?;

        let trimmed_input = input.trim_end();

        if valid.map_or(true, |v| v.contains(&trimmed_input)) {
            break Ok(trimmed_input.to_string());
        }

        println!("\"{}\" is not a valid answer", trimmed_input);

        input.clear();
    }
}

fn main() -> io::Result<()> {
    question("foo", Some(&["bar", "baz"]))?;
    Ok(())
}

File: cs/question.cs
using System;
using System.Collections.Generic;

class Program
{
    // Implementation of the question function in C#
    public static string Question(string prompt, List<string> valid = null)
    {
        string r;

        while (true)
        {
            Console.WriteLine(prompt);
            if (valid != null && valid.Count != 0)
            {
                Console.Write($"({string.Join(", ", valid.ToArray())}): ");
            }
            else
            {
                Console.Write($": ");
            }

            r = Console.ReadLine();

            if (valid != null && valid.Count != 0)
            {
                if (valid.Contains(r))
                {
                    return r;
                }
            }
            else
            {
                return r;
            }

            Console.WriteLine($"\"{r}\" is not a valid answer!");
        }
    }

    public static void Main(string[] args)
    {
        Question("foo", new List<string> { "bar", "baz" });
    }
}
